### PR TITLE
1419 continuous triggers

### PR DESCRIPTION
**Pull request by loadzero**
Merge continuous triggers into core.

### CODE REVIEW

**Comment by usmanm** [comment]
Let's put this under `src/backend/pipeline/trigger` rather than keeping it as an extension.

**Comment by loadzero** [comment]
Agreed. This diff is a straight line port.

**Comment by usmanm** [comment]
Cool, gonna wait for it to get moved then. 

```
/usr/bin/ld: cannot find -lpam
/usr/bin/ld: cannot find -ledit
```

Might want to add them to the requirements in `circle.yaml` and `README.md`.

**Comment by usmanm** [comment]
If it is a missing library or something.

File: bin/bootstrap [hunk 1]
```diff
@@ -42,7 +40,7 @@ def main(args):
       sys.exit(1)
 
     subprocess.call(['pipeline', user, '-c', 'ACTIVATE'])
-    time.sleep(0.1)
+    time.sleep(5)
```
Comment (usmanm) [inline bin/bootstrap]: Why 5 seconds? Seems too high.

File: src/backend/access/transam/xact.c [hunk 2]
```diff
@@ -64,7 +64,6 @@
 #include "utils/timestamp.h"
```
Comment (usmanm) [inline src/backend/access/transam/xact.c]: Revert all changes in this file.

File: src/backend/pipeline/Makefile [hunk 3]
```diff
@@ -20,4 +20,8 @@ OBJS = combinerReceiver.o cont_plan.o update.o stream.o \
 			 fss.o cont_adhoc_format.o dsm_cqueue.o stream_fdw.o cont_execute.o \
 			 adhocReceiver.o cont_adhoc.o transformReceiver.o
 
+SUBDIRS = trigger
+
 include $(top_srcdir)/src/backend/common.mk
+
+# trigger_main
```
Comment (usmanm) [inline src/backend/pipeline/Makefile]: Remove line.

File: bin/bootstrap (hunk 1 above)
Comment (loadzero) [inline bin/bootstrap]: Yeah, that was some debug that slipped in. will fix.

File: src/backend/pipeline/cont_scheduler.c [hunk 4]
```diff
@@ -499,6 +505,10 @@ cont_bgworker_main(Datum arg)
 			am_cont_worker = true;
 			run = &ContinuousQueryWorkerMain;
 			break;
+		case TRIG:
+			is_trigger_process = true;
```
Comment (usmanm) [inline src/backend/pipeline/cont_scheduler.c]: `am_cont_trigger`

File: src/backend/pipeline/trigger/config.c [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+#include "replication/output_plugin.h"
+#include "fmgr.h"
+
+#define PIPELINEDB_TRIGGERS "pipelinedb_triggers"
+
+/* These must be available to pg_dlsym() */
+
+/* guc */
+int alert_socket_mem;
+bool is_trigger_process;
```
Comment (usmanm) [inline src/backend/pipeline/trigger/config.c]: This isn't a guc param. Move this to `cont_scheduler.c` as a static variable and create a helper function `IsContQueryTriggerProcess()`.

File: src/backend/pipeline/trigger/pipeline_triggers--0.9.1.sql [hunk 6]
```diff
@@ -0,0 +1,10 @@
+-- complain if script is sourced in psql, rather than via CREATE EXTENSION
```
Comment (usmanm) [inline src/backend/pipeline/trigger/pipeline_triggers--0.9.1.sql]: This file should be removed.

File: src/backend/pipeline/trigger/pipeline_triggers.control [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# PipelineDB Enterprise extension
```
Comment (usmanm) [inline src/backend/pipeline/trigger/pipeline_triggers.control]: This file should be removed.

File: src/backend/pipeline/trigger/triggerfuncs.c [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+	writev(fd, &iov, 1);
+}
+
+/*
+ * send_alert_new_row
+ *
+ * Converts the 'new' tuple to copy format, and sends to the alert server
+ */
+PG_FUNCTION_INFO_V1(pipeline_send_alert_new_row);
```
Comment (usmanm) [inline src/backend/pipeline/trigger/triggerfuncs.c]: I don't thing we need this declaration here. These functions don't need to be part of the `pg_proc` infrastructure. They should be private functions for the trigger proc.

File: src/backend/utils/fmgr/dfmgr.c [hunk 9]
```diff
@@ -164,7 +164,6 @@ lookup_external_function(void *filehandle, char *funcname)
 	return (PGFunction) pg_dlsym(filehandle, funcname);
 }
 
-
```
Comment (usmanm) [inline src/backend/utils/fmgr/dfmgr.c]: Revert change.

File: src/backend/replication/logical/logical.c [hunk 10]
```diff
@@ -137,7 +137,9 @@ StartupDecodingContext(List *output_plugin_options,
 	 * (re-)load output plugins, so we detect a bad (removed) output plugin
 	 * now.
 	 */
-	LoadOutputPlugin(&ctx->callbacks, NameStr(slot->data.plugin));
+
+	if (strlen(slot->data.plugin.data))
```
Comment (usmanm) [inline src/backend/replication/logical/logical.c]: Why is this change needed? Why would a slot not have a plugin registered? 

`if (strlen(NameStr(slot->data.plugin)))`

File: src/backend/utils/misc/guc.c [hunk 11]
```diff
@@ -2314,7 +2324,7 @@ static struct config_int ConfigureNamesInt[] =
 			NULL
 		},
 		&max_replication_slots,
-		0, 0, MAX_BACKENDS /* XXX? */ ,
+		1, 1, MAX_BACKENDS /* XXX? */ ,
```
Comment (usmanm) [inline src/backend/utils/misc/guc.c]: Wouldn't this cause issues with multiple databases?

File: src/backend/utils/misc/guc.c [hunk 12]
```diff
@@ -2303,7 +2313,7 @@ static struct config_int ConfigureNamesInt[] =
 			NULL
 		},
 		&max_wal_senders,
-		0, 0, MAX_BACKENDS,
+		1, 1, MAX_BACKENDS,
```
Comment (usmanm) [inline src/backend/utils/misc/guc.c]: Wouldn't this cause issues with multiple databases?

File: src/backend/utils/misc/guc.c [hunk 13]
```diff
@@ -3771,7 +3789,7 @@ static struct config_enum ConfigureNamesEnum[] =
 			NULL
 		},
 		&wal_level,
-		WAL_LEVEL_MINIMAL, wal_level_options,
+		WAL_LEVEL_LOGICAL, wal_level_options,
```
Comment (usmanm) [inline src/backend/utils/misc/guc.c]: Let's keep this to `WAL_LEVEL_MINIMAL` and note in the docs that users need to change it to `WAL_LEVEL_LOGICAL` if they want to use triggers. If you're going to do that in a separate pass, that's fine too.

File: src/include/catalog/pg_proc.h [hunk 14]
```diff
@@ -5768,6 +5768,11 @@ DESCR("filtered space saving weighted increment");
 DATA(insert OID = 4417 ( pipeline_exec_adhoc_query PGNSP PGUID 12 1 0 0 0 f f f f t f s 1 0 25 "25" _null_ _null_ _null_ _null_ _null_ pipeline_exec_adhoc_query _null_ _null_ _null_ ));
 DESCR("exec adhoc query");
 
+//DATA(insert OID = 230 (  dsqrt			   PGNSP PGUID 12 1 0 0 0 f f f f t f i 1 0 701 "701" _null_ _null_ _null_ _null_ _null_	dsqrt _null_ _null_ _null_ ));
```
Comment (usmanm) [inline src/include/catalog/pg_proc.h]: Remove commented out stuff.

File: src/include/catalog/pg_proc.h [hunk 15]
```diff
@@ -5768,6 +5768,11 @@ DESCR("filtered space saving weighted increment");
...
+// provolatile
+DATA(insert OID = 4418 ( pipeline_send_alert_new_row PGNSP PGUID 12 1 0 0 0 f f f f t f v 0 0 2279 "" _null_ _null_ _null_ _null_ _null_ pipeline_send_alert_new_row _null_ _null_ _null_ ));
```
Comment (usmanm) [inline src/include/catalog/pg_proc.h]: I feel like this function shouldn't be exposed. It's an internal function of the alert server.

File: src/include/pipeline/cont_scheduler.h [hunk 16]
```diff
@@ -34,6 +34,7 @@ typedef enum
 	COMBINER,
 	WORKER,
 	ADHOC,
+	TRIG,
```
Comment (usmanm) [inline src/include/pipeline/cont_scheduler.h]: `TRIGGER`

File: src/include/pipeline/cont_scheduler.h [hunk 17]
```diff
@@ -71,6 +72,7 @@ struct ContQueryDatabaseMetadata
 	int adhoc_counter;
 	/* Number of entries is equal to max_worker_processes. */
 	ContQueryProc *adhoc_procs;
+	ContQueryProc *trigger_proc;
```
Comment (usmanm) [inline src/include/pipeline/cont_scheduler.h]: Since there's always only one processes, you can just embed the full `struct`.

File: src/test/py/base.py [hunk 18]
```diff
@@ -267,7 +344,7 @@ def clean_db(request):
     """
     pdb = request.module.pipeline
     request.addfinalizer(pdb.drop_all_queries)
-
+    pdb.execute("select pipeline_trigger_debug('setup')")
```
Comment (usmanm) [inline src/test/py/base.py]: What does this do?

File: src/backend/replication/logical/logical.c (hunk 10 above)
Comment (loadzero) [inline src/backend/replication/logical/logical.c]: The plugin is not dynamically linked any more. We should probably replace this hack with a hook.

Comment (loadzero) [inline src/backend/replication/logical/logical.c]: The plugin is not dynamically linked any more. We should probably replace this hack with a hook.

File: src/include/pipeline/cont_scheduler.h (hunk 16 above)
Comment (loadzero) [inline src/include/pipeline/cont_scheduler.h]: conflicts with another enum

File: src/test/py/base.py (hunk 18 above)
Comment (loadzero) [inline src/test/py/base.py]: This is a replacement for trigger_testing_setup - it clears the shared file that is being written to by test_alert_new_row. I will move it into a specific fixture.

File: src/include/catalog/pg_proc.h (hunk 15 above)
Comment (loadzero) [inline src/include/catalog/pg_proc.h]: pipeline_send_alert_new_row is the trigger function that enables sending alerts from a trigger, it currently needs to be public to work with the current syntax

e.g:

```
create continuous view v as select x::text,count(*) from s group by x;

create trigger trigg after update or insert on v for row 
WHEN (true)
execute procedure pipeline_send_alert_new_row();
```

File: src/backend/pipeline/trigger/trigger.c [hunk 19]
```diff
@@ -0,0 +1,1181 @@
...
+	destroy_trigger_process_state(state);
+	destroy_wal_stream(ws);
+	proc_exit(0);
+}
+
+/*
+ * RegisterTriggerProcess
+ */
+void
+RegisterTriggerProcess(void)
```
Comment (usmanm) [inline src/backend/pipeline/trigger/trigger.c]: Remove function.

File: src/backend/pipeline/trigger/trigger.c [hunk 20]
```diff
@@ -0,0 +1,1181 @@
...
+	state->cur_wal_batch = NULL;
+}
+
+/*
+ * ResetTriggerCacheEntry
+ */
+static void
+ResetTriggerCacheEntry(TriggerProcessState *state, TriggerCacheEntry *entry)
+{
+	/* XXX - this currently leaks */
```
Comment (usmanm) [inline src/backend/pipeline/trigger/trigger.c]: If it does, let's create an issue. Otherwise, remove comment.

File: src/backend/pipeline/trigger/trigger.c [hunk 21]
```diff
@@ -0,0 +1,1181 @@
...
+	cl = get_changelist(state, xact_batch, relid);
+
+	if (action == TRIGGER_PROCESS_CHANGE_DELETE ||
+	 	action == TRIGGER_PROCESS_CHANGE_NOOP)
+		return;
+
+	if (!entry->numtriggers)
+		return;
+
+	// TRIGGER
```
Comment (usmanm) [inline src/backend/pipeline/trigger/trigger.c]: What does this comment mean? Use /* */ style comments.

File: src/backend/replication/logical/logical.c (hunk 10 above)
Comment (usmanm) [inline src/backend/replication/logical/logical.c]: Sounds good, just put a comment above it and make the `NameStr` change.

File: src/include/catalog/pg_proc.h (hunk 15 above)
Comment (usmanm) [inline src/include/catalog/pg_proc.h]: Ah! Makes sense.

File: src/include/pipeline/cont_scheduler.h (hunk 16 above)
Comment (usmanm) [inline src/include/pipeline/cont_scheduler.h]: Gross. `TRIG` is cool then.

File: src/test/py/base.py (hunk 18 above)
Comment (usmanm) [inline src/test/py/base.py]: Let's create a finalizer function for it then, like `drop_all_queries`.

File: src/backend/pipeline/trigger/trigger.c (hunk 21 above)
Comment (loadzero) [inline src/backend/pipeline/trigger/trigger.c]: Yeah, if you see any // comments from me, they are almost always placeholders that have been erroneously leftover. I usually grep for them before committing to master.

Comment (usmanm) [inline src/backend/pipeline/trigger/trigger.c]: Sounds good!

**Comment by usmanm** [comment]
We should also remove the trigger proc's replication slot when a database is dropped.
